src/server/service/diagnostics.rs

```diff
@@ -0,0 +1,684 @@
+    }
+
+    #[derive(Debug)]
+    pub enum Error {
+        InvalidRequest(String),
+        ParseError(String),
+        SearchError(String),
+        IOError(std::io::Error),
+    }
+
+    impl From<std::io::Error> for Error {
+        fn from(err: std::io::Error) -> Self {
+            Error::IOError(err)
+        }
+    }
+
+    impl LogIterator {
+        fn new<P: AsRef<Path>>(
+            log_file: P,
+            begin_time: i64,
+            end_time: i64,
+            level: LogLevel,
+            filter: String,
+        ) -> Result<Self, Error> {
+            let log_path = log_file.as_ref();
+            let log_name = match log_path.file_name() {
+                Some(file_name) => match file_name.to_str() {
+                    Some(file_name) => file_name,
+                    None => {
+                        return Err(Error::SearchError(format!("Invalid utf8: {:?}", file_name)))
+                    }
+                },
+                None => {
+                    return Err(Error::SearchError(format!(
+                        "Illegal file name: {:?}",
+                        log_path
+                    )))
+                }
+            };
+            if log_name.is_empty() {
+                return Err(Error::InvalidRequest("Empty `log_file` path".to_owned()));
+            }
+            let log_dir = match log_path.parent() {
+                Some(dir) => dir,
+                None => {
+                    return Err(Error::SearchError(format!(
+                        "Illegal parent dir: {:?}",
+                        log_path
+                    )))
+                }
+            };
+
+            let mut search_files = vec![];
+            for entry in read_dir(log_dir)? {
+                let entry = entry?;
+                if !entry.path().is_file() {
+                    continue;
+                }
+                let file_name = entry.file_name();
+                let file_name = match file_name.to_str() {
+                    Some(file_name) => file_name,
+                    None => continue,
+                };
+                // Rotated file name have the same prefix with the original
+                if !file_name.starts_with(log_name) {
+                    continue;
+                }
+                // Open the file
+                let mut file = match File::open(entry.path()) {
+                    Ok(file) => file,
+                    Err(_) => continue,
+                };
+
+                let (file_start_time, file_end_time) = match parse_time_range(&file) {
+                    Ok((file_start_time, file_end_time)) => (file_start_time, file_end_time),
+                    Err(_) => continue,
+                };
+                if (begin_time <= file_start_time && end_time >= file_start_time)
+                    || (end_time >= file_end_time && begin_time <= file_end_time)
+                {
+                    if let Err(err) = file.seek(SeekFrom::Start(0)) {
```

Comment:
It's unnecessary.

Reply:
It' necessary because the same file descriptor will be reused.